### PR TITLE
Plan to Torchscript translation

### PR DESCRIPTION
**Pull request by vvmnnnkv**
## Description
Part of https://github.com/OpenMined/PySyft/issues/5045:
1. Defines Plan with basic MLP model (backward pass is coded by hand)
2. Translate this Plan to torchscript
3. Torchscript produces same output as syft Plan
Note:
4. Need to test torchscript on mobiles

## Affected Dependencies
Added syft_proto dependency to syft, to use old Plan protobuf

## How has this been tested?
- Unit test

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

**Review by cereallarceny — APPROVED** [review]

File: src/syft/core/plan/translation/torchscript/plan_translate.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            self._kwarg_names = kwarg_names
+
+        def forward(self, *args: Any) -> Any:
+            kwarg_ptrs: Dict[str, Pointer] = {}
+
+            # Since Syft Plan needs pointers as args,
+            # reverse-map actual arg values to pointers
+            for name, arg in zip(self._kwarg_names, args):
+                found = False
+                for ptr in ROOT_CLIENT.store:
```
Comment (koenvanderveen) [inline src/syft/core/plan/translation/torchscript/plan_translate.py:45]: It would be great if we could somehow simplify this search. The whole part of Actions working with pointers is also somewhat annoying in the normal `Plan` class. Maybe for now extract this complexity in a function/classmethod to improve readability?

**Review by koenvanderveen — APPROVED** [review]

File: src/syft/core/plan/translation/torchscript/plan_translate.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        so we need to have kwarg names inside torchscript to pass them into Plan.
+        We use nn.Module to store input kwarg names inside torchscript.
+        """
+
+        def __init__(self, kwarg_names: List[str]):
+            _kwarg_names: List[str]
+            super(PlanWrapper, self).__init__()
+            self._kwarg_names = kwarg_names
+
+        def forward(self, *args: Any) -> Any:
```
Comment (koenvanderveen) [inline src/syft/core/plan/translation/torchscript/plan_translate.py:38]: maybe execute is a better name, as people may associate `.forward` with `torch.nn.module.forward`

**Comment by koenvanderveen** [comment]
In general: looks great!

File: src/syft/core/plan/translation/torchscript/plan_translate.py (hunk 2 above)
Comment (vvmnnnkv) [reply src/syft/core/plan/translation/torchscript/plan_translate.py:38]: it really is nn.Module :) 
This wrapper is a workaround to store extra state (kwarg names for Plan) that can't be passed as argument into torchscript (note: it's possible to pass dict to torchscript, but dict values need to have same type, i.e. you'd need to flatten model params list into this dict)

File: src/syft/core/plan/translation/torchscript/plan_translate.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            # reverse-map actual arg values to pointers
+            for name, arg in zip(self._kwarg_names, args):
+                found = False
+                for ptr in ROOT_CLIENT.store:
+                    store_value: Any = PLAN_BUILDER_VM.store[ptr.id_at_location].data
+                    is_list = isinstance(arg, (SyList, list)) and isinstance(
+                        store_value, (SyList, list)
+                    )
+                    is_dict = isinstance(
+                        arg, (dict, Dict, OrderedDict, SyOrderedDict)
```
Comment (tudorcebere) [inline src/syft/core/plan/translation/torchscript/plan_translate.py:51]: could we define some constants somewhere? Like: LIST_TYPES = (SyList, list), DICT_TYPES = (dict, Dict, OrderedDict, SyOrderedDict)? you can store these in syft/lib/python?

File: src/syft/core/plan/translation/torchscript/plan_translate.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    # Builder VM holds inputs in the store, retrieve actual arg values from there
+    kwarg_names = list(plan.inputs.keys())
+    args = tuple(
+        PLAN_BUILDER_VM.store[plan.inputs[name].id_at_location].data
+        for name in kwarg_names
+    )
+
+    # Remove wrappers, th.jit.trace doesn't like them
+    args = tuple(list(arg) if isinstance(arg, SyList) else arg for arg in args)
```
Comment (tudorcebere) [inline src/syft/core/plan/translation/torchscript/plan_translate.py:93]: I think you can check for SyPrimitive and then use .downcast() on it? note sure.

File: src/syft/core/plan/translation/torchscript/plan.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        As a part of serialization and deserialization, we need the ability to
+        lookup the protobuf object type directly from the object type. This
+        static method allows us to do this.
+
+        Importantly, this method is also used to create the reverse lookup ability within
+        the metaclass of Serializable. In the metaclass, it calls this method and then
+        it takes whatever type is returned from this method and adds an attribute to it
+        with the type of this class attached to it. See the MetaSerializable class for details.
+
+        :return: the type of protobuf object which corresponds to this class.
```
Comment (jmaunon) [inline src/syft/core/plan/translation/torchscript/plan.py:38]: Please, Use Google convention

**Review by jmaunon — COMMENTED** [review]
@vvmnnnkv , Please update docstring to Google convention (see https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

File: src/syft/core/plan/translation/torchscript/plan_translate.py (hunk 2 above)
Comment (koenvanderveen) [reply src/syft/core/plan/translation/torchscript/plan_translate.py:38]: Oh wow, it took me a few minutes but now I get it. I would document this extremely clearly, not just in the init but also in the forward method, because it makes sense once you get it but I think it could potentially confuse many people.

File: src/syft/core/plan/translation/torchscript/plan_translate.py (hunk 3 above)
Comment (madhavajay) [reply src/syft/core/plan/translation/torchscript/plan_translate.py:51]: Yeah it would be nice to have Unions of these be auto generated for these kinds of checks.

**Review by cereallarceny — APPROVED** [review]

File: src/syft/core/plan/translation/torchscript/plan_translate.py (hunk 3 above)
Comment (juharris) [reply src/syft/core/plan/translation/torchscript/plan_translate.py:51]: `OrderedDict` is a `dict` so you don't really need it in the list:
```python
from collections import OrderedDict
isinstance(OrderedDict(), dict)
Out[8]: True
```

`SyOrderedDict` is an `OrderedDict` so it's not really needed either unless you think that might changed in the future but I assume it will always at least be a `dict`.

Looking around a bit online, it looks like you don't really need `Dict` in there either: https://stackoverflow.com/questions/62547848/should-isinstance-check-against-typing-or-collections-abc

so `isinstance(arg, dict)` should be sufficient.
